### PR TITLE
Landmark recognition

### PR DESCRIPTION
**Pull request by tzufgoogle**
Bert-based entity recognition model for identifying landmarks in sentences. Input: sentence. Output: for each token whether it is a landmark or not. The result is ~94% accuracy and ~80 f1 score. 

To run this do:
bazel-bin/cabby/model/landmark_recognition/entity_recognition \
  --data_dir ~/data/RVS/Manhattan  \

If you want to test samples from Touchdown\RUN then also add their paths (data_dir_touchdown\data_dir_run)
The samples will be printed and the landmark recognized will appear in green.

### CODE REVIEW

File: cabby/model/landmark_recognition/dataset.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      if len(spans)>0:
+        span = spans.pop(0)
+        start, end = span[0], span[1]
+      else:
+        start = len(sentence)+1
+
+  tokenized_input = tokenize_and_align_labels(labels, sentence_words)
+  return tokenized_input
+
+def tokenize_and_align_labels(tags, sentence_word, label_all_tokens = True):
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset.py:130]: Documentation please! Example inputs and outputs much appreciated.

File: cabby/model/landmark_recognition/dataset.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      label_ids.append(tags[word_idx] if label_all_tokens else -100)
+    previous_word_idx = word_idx
+
+  assert len(label_ids) == len(tokenized_inputs['input_ids'])
+  tokenized_inputs["labels"] = label_ids
+  return tokenized_inputs
+
+
+
+class PadSequence:
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset.py:155]: This class has 4-space tabs instead of 2.

Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset.py:155]: Why is this a class and not just a function pad_sequence(batch)?

File: cabby/model/landmark_recognition/dataset_bert.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    data_dir: The directory of the data.
+    region: The region of the data.
+    s2level: The s2level of the cells.
+  Returns:
+    The train, validate and test sets.
+  '''
+  rvs_dataset = datasets.RVSDataset(data_dir, s2level, region)
+
+  train_dataset = EntityRecognitionSplit(
+    rvs_dataset.train)
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset_bert.py:73]: Extraneous new lines in this block, like the other one.

File: cabby/model/landmark_recognition/dataset_bert.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    cur_index += len(word)
+    if cur_index<len(sentence) and sentence[cur_index] == ' ':
+      cur_index+=1
+    if cur_index >= end:
+      if len(spans)>0:
+        span = spans.pop(0)
+        start, end = span[0], span[1]
+      else:
+        start = len(sentence)+1
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset_bert.py:123]: This looks like a similar loop from before (maybe exact duplicate?). Can you abstract this in a function and write tests for it?

File: cabby/model/landmark_recognition/dataset_bert.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+      label_ids.append(tags[word_idx])
+    previous_word_idx = word_idx
+
+  assert len(label_ids) == len(tokenized_inputs['input_ids'])
+
+  tokenized_inputs["labels"] = label_ids
+  return tokenized_inputs
+
+
+class PadSequence:
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset_bert.py:162]: Actually, there is even more (apparently) duplicated with dataset.py. Can you refactor so there isn't this duplication?

File: cabby/model/landmark_recognition/dataset_er.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+from transformers import DistilBertTokenizerFast
+from transformers import AutoModelForTokenClassification, AutoTokenizer
+# from keras.preprocessing.sequence import pad_sequences
+from nltk.tokenize import word_tokenize
+
+
+tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
+MAX_LEN = 256
+
+
```
Comment (jasonbaldridge) [inline cabby/model/landmark_recognition/dataset_er.py:38]: Lots of duplication with other dataset*.py  files before.  Please refactor.

**Review by jasonbaldridge — COMMENTED** [review]
Looks promising! There's lots of code here, so I tried to scan through it all, but may have missed some things. Main things are:

- please add more to the description of this PR, including some of the command line calls used to create the data and run the model (even better: add them to run_all.sh or something like it)
- There is a bunch of duplicated code in here, so it would be good to refactor and reduce redundancy.

Thanks!

File: cabby/model/landmark_recognition/dataset.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    cur_index += len(word)
+    if cur_index<len(sentence) and sentence[cur_index] == ' ':
+      cur_index+=1
+    if cur_index >= end:
+      if len(spans)>0:
+        span = spans.pop(0)
+        start, end = span[0], span[1]
+      else:
+        start = len(sentence)+1
```
[the remark this answers is not in the document]
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset.py]: Added documentation.

File: cabby/model/landmark_recognition/dataset.py (hunk 2 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset.py]: I will remove dataset.py and this will be fixed.

Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset.py]: This is the Pytorch way.

File: cabby/model/landmark_recognition/dataset.py (hunk 1 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset.py]: There is documentation in the dataset_bert.py. I will remove dataset.py

File: cabby/model/landmark_recognition/dataset_bert.py (hunk 3 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset_bert.py:73]: Fixed.

File: cabby/model/landmark_recognition/dataset_bert.py (hunk 4 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset_bert.py:123]: I will remove dataset.py

File: cabby/model/landmark_recognition/dataset_bert.py (hunk 5 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset_bert.py:162]: I will remove dataset.py

File: cabby/model/landmark_recognition/dataset_er.py (hunk 6 above)
Comment (tzufgoogle) [reply cabby/model/landmark_recognition/dataset_er.py:38]: I will remove dataset.py

**Review by tzufgoogle — COMMENTED** [review]
The dataset.py file name was changed to datase_bert.py but both files appeared in the git. I removed the dataset.py file.

File: cabby/model/landmark_recognition/dataset.py (hunk 2 above)
Comment (jasonbaldridge) [reply cabby/model/landmark_recognition/dataset.py]: Okay!

**Review by jasonbaldridge — APPROVED** [review]
LGTM. Please do add a few command line examples to the PR description too. Thanks!
